python/lsst/daf/butler/script/queryDatasets.py

```diff
@@ -0,0 +1,184 @@
+    """Get dataset refs from a repository.
+
+    Parameters
+    ----------
+    repo : `str`
+        URI to the location of the repo or URI to a config file describing the
+        repo and its location.
+    glob : iterable [`str`]
+        A list of glob-style search string that fully or partially identify
+        the dataset type names to search for.
+    collections : iterable [`str`]
+        A list of glob-style search string that fully or partially identify
+        the collections to search for.
+    where : `str`
+        A string expression similar to a SQL WHERE clause.  May involve any
+        column of a dimension table or (as a shortcut for the primary key
+        column of a dimension table) dimension name.
+    find_first : `bool`
+        For each result data ID, only yield one DatasetRef of each DatasetType,
+        from the first collection in which a dataset of that dataset type
+        appears (according to the order of `collections` passed in).  If used,
+        `collections` must specify at least one expression and must not contain
+        wildcards.
+    show_uri : `bool`
+        If True, include the dataset URI in the output.
+    Returns
+    -------
+    datasetTables : `list` [``astropy.table._Table``]
+        A list of astropy tables, one for each dataset type.
+    """
+    butler = Butler(repo)
+
+    dataset = globToRegex(glob)
+    if not dataset:
+        dataset = ...
+
+    if collections and not find_first:
+        collections = globToRegex(collections)
+    elif not collections:
+        collections = ...
+
+    datasets = butler.registry.queryDatasets(datasetType=dataset,
+                                             collections=collections,
+                                             where=where,
+                                             deduplicate=find_first)
+
+    tables = _Tables()
+
+    def addDatasetComponent(datasetRef, name, uri):
+        """Add an entry in the correct output dataset table for the given
+        DatasetRef and component name, and component URI."""
+        datasetTypeName = datasetRef.datasetType.name
+        if name:
+            datasetTypeName = ".".join((datasetTypeName, name))
+        table = tables.getTable(datasetTypeName)
+        if not table:
+            columnNames = ["type", "run", "id", *[str(item) for item in datasetRef.dataId.keys()]]
+            if show_uri:
+                columnNames.append("URI")
+            table = tables.makeTable(datasetTypeName, columnNames)
+        extraValues = [str(value) for value in datasetRef.dataId.values()]
+        if show_uri:
+            extraValues += [uri]
+        table.addRow((datasetTypeName,
+                      datasetRef.run,
+                      datasetRef.id,
+                      *extraValues))
+
+    for datasetRef in datasets:
+        primaryURI, componentURIs = butler.getURIs(datasetRef, collections=datasetRef.run)
```

Comment:
getURIs takes a non-zero amount of time so I wouldn't call it unless `show_uri` is `True`.

Reply:
As far as I've been able to tell, getURIs is the only way to get the components of the datasetRef. Is there another way, or do we want to *not* show the components if `show_uri` is false?

Reply:
If people explicitly ask for `raw.wcs` as a dataset type then they can get the components they ask for and getURIs should do the right thing.  getURIs will only return multiple values if the dataset has been disassembled (in which case each component gets a separate URI -- if it has not been disassembled each component would get the same URI). Therefore getURIs is only useful if people want to see URIs. If they don't want URIs don't call it at all and give people what they got from queryDatasets (which may or may not include a component name)